File: orchestrator/boards.py
"""Kart (fiziksel PCB) katmani — tek dogruluk kaynagi.

Kart, elektriksel modele DIK bir konum katmanidir: cihazin attach.controller_id /
via_mux alanlari degismez, board_id yalnizca fiziksel yeri soyler. `boards`
tanimli degilse sistem tek ortuk ana karttan ibaret sayilir ve uretilen cikti
bugunku duzeninde kalir (bkz. docs/superpowers/specs/2026-08-04-multi-board-topology-design.md).
"""
from __future__ import annotations

import re

MAIN_BOARD_ID = "main"

#: Turkce harfler ASCII karsiliklarina katlanir (C tanimlayici uretimi icin).
_FOLD = {
    "ı": "i", "İ": "I", "ş": "s", "Ş": "S", "ğ": "g", "Ğ": "G",
    "ü": "u", "Ü": "U", "ö": "o", "Ö": "O", "ç": "c", "Ç": "C",
}


def _fold_ascii(text: str) -> str:
    return "".join(_FOLD.get(ch, ch) for ch in text)


def _words(name: str) -> list[str]:
    folded = _fold_ascii(name or "")
    return [w for w in re.split(r"[^A-Za-z0-9]+", folded) if w]


def board_identifier(name: str) -> str:
    """Kart adindan C camelCase tanimlayici: "RF Kart" -> "rfKart"."""
    words = _words(name)
    if not words:
        return "kart"
    head = words[0].lower()
    tail = "".join(w[:1].upper() + w[1:].lower() for w in words[1:])
    identifier = head + tail
    if not identifier[0].isalpha():
        identifier = "k" + identifier
    return identifier


def board_dirname(name: str) -> str:
    """Kart adindan snake_case klasor adi: "RF Kart" -> "rf_kart"."""
    words = _words(name)
    if not words:
        return "kart"
    dirname = "_".join(w.lower() for w in words)
    if not dirname[0].isalpha():
        dirname = "k_" + dirname
    return dirname


def assert_unique_identifiers(board_list: list[dict]) -> None:
    """Iki kart ayni C tanimlayiciya duserse sessiz cakisma yerine acik hata."""
    seen: dict[str, str] = {}
    for board in board_list:
        identifier = board_identifier(str(board.get("name", "")))
        previous = seen.get(identifier)
        if previous is not None:
            raise ValueError(
                f"kart adlari ayni C tanimlayiciya dusuyor ('{identifier}'): "
                f"'{previous}' ve '{board.get('name')}' — kart adlarini ayirin")
        seen[identifier] = str(board.get("name", ""))


def boards_declared(spec: dict) -> bool:
    """Kullanici kart tanimladi mi? TUM 'kart modu' kararlari buna bakar."""
    return bool(spec.get("boards"))


def normalized_boards(spec: dict) -> list[dict]:
    """Kart listesi; tanimli degilse tek ortuk ana kart dondurur."""
    declared = spec.get("boards") or []
    if declared:
        return [{**board, "implicit": False} for board in declared]
    name = str((spec.get("project") or {}).get("name") or "Ana Kart")
    return [{"id": MAIN_BOARD_ID, "name": name, "role": "main", "implicit": True}]


def board_id_of(entity: dict) -> str:
    """Cihaz/mux'un kart kimligi; verilmemisse ana kart."""
    return str(entity.get("board_id") or MAIN_BOARD_ID)


def main_board_id(spec: dict) -> str:
    for board in normalized_boards(spec):
        if str(board.get("role")) == "main":
            return str(board["id"])
    return MAIN_BOARD_ID


def part_slug(part: str) -> str:
    return re.sub(r"[^a-z0-9]", "", str(part).lower()) or "cihaz"


def normalize_device_ids(spec: dict) -> dict:
    """Cihaz/mux kimliklerini `<kart>_<parca>[_<n>]` kuralina ceker (frontend
    lib/boards.ts::normalizeDeviceIds ile AYNI kural); yeni spec kopyasi dondurur.

    Kart oneki kart adinin snake_case'i (kart tanimsizsa "kart"); ayni kartta ayni parcadan
    birden fazla cihaz varsa ekleme sirasiyla _1, _2 ... (tek ise sonek yok).
    """
    board_list = normalized_boards(spec)
    declared = bool(spec.get("boards"))
    by_id = {str(b["id"]): b for b in board_list}
    main_id = main_board_id(spec)

    def prefix_of(item: dict) -> str:
        if not declared:
            return "kart"
        bid = str(item.get("board_id") or main_id)
        board = by_id.get(bid) or by_id.get(main_id) or board_list[0]
        return board_dirname(str(board["name"]))

    def assign(items: list[dict]) -> dict[str, str]:
        counts: dict[str, int] = {}
        for item in items:
            key = f"{prefix_of(item)}_{part_slug(item.get('part', ''))}"
            counts[key] = counts.get(key, 0) + 1
        seen: dict[str, int] = {}
        mapping: dict[str, str] = {}
        for item in items:
            key = f"{prefix_of(item)}_{part_slug(item.get('part', ''))}"
            seen[key] = seen.get(key, 0) + 1
            mapping[str(item.get("id"))] = f"{key}_{seen[key]}" if counts[key] > 1 else key
        return mapping

    devices = [dict(d) for d in spec.get("devices", [])]
    muxes = [dict(m) for m in spec.get("muxes", [])]
    dev_map, mux_map = assign(devices), assign(muxes)
    for d in devices:
        d["id"] = dev_map.get(str(d.get("id")), d.get("id"))
        via = (d.get("attach") or {}).get("via_mux")
        if isinstance(via, dict) and via.get("mux_id") in mux_map:
            d["attach"] = {**d["attach"], "via_mux": {**via, "mux_id": mux_map[via["mux_id"]]}}
    for m in muxes:
        m["id"] = mux_map.get(str(m.get("id")), m.get("id"))
    return {**spec, "devices": devices, "muxes": muxes}

File: frontend/src/lib/boards.ts
/** Kart (fiziksel PCB) katmani — UI tarafindaki tek dogruluk kaynagi.
 *
 * Kurallar `orchestrator/boards.py` ile AYNI olmak zorundadir: burada uretilen
 * kart kimligi spec'e yazilir, uretim tarafi ondan klasor adi ve C tanimlayici
 * turetir. Kart tanimli degilse (boards bos) sistem tek ortuk ana karttir ve
 * kanvas bugunku duzeninde kalir.
 */
import type { Board } from "@/lib/types";

export const MAIN_BOARD_ID = "main";

/** React Flow dugum kimligi cihaz kimlikleriyle carpismasin diye onek alir. */
export const BOARD_NODE_PREFIX = "board-";

/** Turkce harfler ASCII karsiliklarina katlanir (boards.py::_FOLD ile ayni). */
const TR_FOLD: Record<string, string> = {
  "ı": "i", "İ": "i", "ş": "s", "Ş": "s", "ğ": "g", "Ğ": "g",
  "ü": "u", "Ü": "u", "ö": "o", "Ö": "o", "ç": "c", "Ç": "c",
};

function foldAscii(text: string): string {
  return Array.from(text).map((ch) => TR_FOLD[ch] ?? ch).join("");
}

function words(name: string): string[] {
  return foldAscii(name ?? "").split(/[^A-Za-z0-9]+/).filter(Boolean);
}

/** Kart/konnektor adindan snake_case kimlik: "RF Kart" -> "rf_kart".
 *  Sema deseni: ^[a-z][a-z0-9_]*$ */
export function boardSlug(name: string, fallback: string): string {
  const parts = words(name);
  let id = parts.map((w) => w.toLowerCase()).join("_");
  if (!id) id = fallback;
  if (!/^[a-z]/.test(id)) id = `k_${id}`;
  return id;
}

/** Ayni kimlik iki kez uretilmesin: "rf_kart", "rf_kart_2", ... */
export function uniqueId(base: string, taken: Iterable<string>): string {
  const used = new Set(taken);
  if (!used.has(base)) return base;
  let n = 2;
  while (used.has(`${base}_${n}`)) n += 1;
  return `${base}_${n}`;
}

/** Ana kartin kimligi; kart yoksa ortuk "main". */
export function mainBoardId(boards: Board[]): string {
  return (boards.find((b) => b.role === "main") ?? boards[0])?.id ?? MAIN_BOARD_ID;
}

/** Cihaz/mux'un kart kimligi. Tanimsiz kimlik ANA KARTA duser — codegen'deki
 *  `_effective_board_id` ile ayni davranis (cihaz sessizce kaybolmaz). */
export function effectiveBoardId(entity: { board_id?: string }, boards: Board[]): string {
  const main = mainBoardId(boards);
  const bid = entity.board_id;
  if (!bid) return main;
  return boards.some((b) => b.id === bid) ? bid : main;
}

export function boardNodeId(boardId: string): string {
  return `${BOARD_NODE_PREFIX}${boardId}`;
}

/** React Flow dugum kimliginden kart kimligi; kart dugumu degilse null. */
export function boardIdFromNode(nodeId: string | null | undefined): string | null {
  if (!nodeId || !nodeId.startsWith(BOARD_NODE_PREFIX)) return null;
  return nodeId.slice(BOARD_NODE_PREFIX.length);
}

export interface BoardGroup<T> {
  boardId: string;
  boardName: string;
  items: T[];
}

/** Ogeleri kart kimligine gore gruplar (CIT/Test Bench panelleri — Task 4).
 *  Grup sirasi `boardList` sirasini izler (manifest.boards sirasi = kullanicinin
 *  ekleme sirasi, ana kart en basta gelir — bkz. store addBoard). `boardList`'te
 *  TANIMLI OLMAYAN bir board_id (beklenmez ama dogrulayici atlanmis/eski
 *  manifest olabilir) sessizce KAYBOLMAZ: kendi id'siyle ayri bir grup olarak,
 *  ilk gorulme sirasiyla sona eklenir. */
export function groupByBoardId<T>(
  items: T[],
  boardIdOf: (item: T) => string,
  boardList: Array<{ id: string; name: string }>,
): BoardGroup<T>[] {
  const byId = new Map<string, T[]>();
  for (const item of items) {
    const id = boardIdOf(item);
    const bucket = byId.get(id);
    if (bucket) bucket.push(item);
    else byId.set(id, [item]);
  }
  const groups: BoardGroup<T>[] = [];
  for (const board of boardList) {
    const bucket = byId.get(board.id);
    if (bucket) {
      groups.push({ boardId: board.id, boardName: board.name, items: bucket });
      byId.delete(board.id);
    }
  }
  for (const [id, bucket] of byId) {
    groups.push({ boardId: id, boardName: id, items: bucket });
  }
  return groups;
}

/** Cihaz/mux kimligi kurali (kullanici istegi 2026-09-06): `<kart>_<parca>[_<n>]`.
 *  Kart oneki kart ADININ slug'i (kart tanimsizsa "kart"); ayni kartta ayni parcadan birden
 *  fazla cihaz varsa ekleme sirasiyla `_1, _2, ...` (tek ise sonek yok): sakk_ltc2991_1.
 *  Uretilen enum (`I2C_CIHAZ_SAKK_LTC2991_1`), CIT varsayilan adlari ve ekranlar bu kimligi
 *  kullanir. `orchestrator/boards.py::normalize_device_ids` ile AYNI kural. */
export function partSlug(part: string): string {
  return part.toLowerCase().replace(/[^a-z0-9]/g, "") || "cihaz";
}

interface IdItem {
  id: string;
  part: string;
  board_id?: string;
}

function assignIds<T extends IdItem>(items: T[], boards: Board[]): Map<string, string> {
  const prefixOf = (item: T): string => {
    if (!boards.length) return "kart";
    const bid = effectiveBoardId(item, boards);
    const board = boards.find((b) => b.id === bid);
    return board ? boardSlug(board.name, board.id) : bid;
  };
  const counts = new Map<string, number>();
  for (const item of items) {
    const key = `${prefixOf(item)}_${partSlug(item.part)}`;
    counts.set(key, (counts.get(key) ?? 0) + 1);
  }
  const seen = new Map<string, number>();
  const mapping = new Map<string, string>();
  for (const item of items) {
    const key = `${prefixOf(item)}_${partSlug(item.part)}`;
    const n = (seen.get(key) ?? 0) + 1;
    seen.set(key, n);
    mapping.set(item.id, (counts.get(key) ?? 1) > 1 ? `${key}_${n}` : key);
  }
  return mapping;
}

export interface NormalizedIds<D, M> {
  devices: D[];
  muxes: M[];
  deviceMap: Map<string, string>;
  muxMap: Map<string, string>;
}

/** Kimlikleri kurala gore yeniden adlandirir; degisiklik yoksa null. Mux referanslari
 *  (attach.via_mux.mux_id) da yeni kimlige tasinir. */
export function normalizeDeviceIds<
  D extends IdItem & { attach: { via_mux?: { mux_id: string; channel: number } | null } },
  M extends IdItem,
>(devices: D[], muxes: M[], boards: Board[]): NormalizedIds<D, M> | null {
  const deviceMap = assignIds(devices, boards);
  const muxMap = assignIds(muxes, boards);
  const changed =
    [...deviceMap].some(([from, to]) => from !== to) || [...muxMap].some(([from, to]) => from !== to);
  if (!changed) return null;
  const newDevices = devices.map((d) => {
    const via = d.attach.via_mux;
    const attach = via ? { ...d.attach, via_mux: { ...via, mux_id: muxMap.get(via.mux_id) ?? via.mux_id } } : d.attach;
    return { ...d, id: deviceMap.get(d.id) ?? d.id, attach };
  });
  const newMuxes = muxes.map((m) => ({ ...m, id: muxMap.get(m.id) ?? m.id }));
  return { devices: newDevices, muxes: newMuxes, deviceMap, muxMap };
}
